pkg/sql/util/composite_primary_key_util.go

```diff
@@ -0,0 +1,243 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"github.com/fagongzi/util/format"
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+	"strconv"
+)
+
+var prefixPriColName string = "mo_cpkey_"
+
+func ExtractCompositePrimaryKeyColumnFromColDefs(colDefs []*plan.ColDef) ([]*plan.ColDef, *plan.ColDef) {
+	for num := range colDefs {
+		if colDefs[num].IsCPkey {
+			cPKC := colDefs[num]
+			colDefs = append(colDefs[:num], colDefs[num+1:]...)
+			return colDefs, cPKC
+		}
+	}
+	return colDefs, nil
+}
+
+func BuildCompositePrimaryKeyColumnName(s []string) string {
+	var name string
+	name = prefixPriColName
+	for _, single := range s {
+		lenNum := format.Int64ToString(int64(len(single)))
+		for num := 0; num < 3-len(lenNum); num++ {
+			name += string('0')
+		}
+		name += lenNum
+		name += single
+	}
+	return name
+}
+
+func SplitCompositePrimaryKeyColumnName(s string) []string {
+	var names []string
+	for next := len(prefixPriColName); next < len(s); {
+		strLen, _ := strconv.Atoi(s[next : next+3])
+		names = append(names, s[next+3:next+3+strLen])
+		next += strLen + 3
+	}
+
+	return names
+}
+
+func FillCompositePKeyBatch(bat *batch.Batch, p *plan.ColDef, rowCount int) error {
+	names := SplitCompositePrimaryKeyColumnName(p.Name)
+	cPkeyVecMap := make(map[string]*vector.Vector)
+	for num, attrName := range bat.Attrs {
+		for _, name := range names {
+			if attrName == name {
+				cPkeyVecMap[name] = bat.Vecs[num]
+			}
+		}
+	}
+	vs := make([][]byte, rowCount)
+	ps := types.NewPackerArray(rowCount)
+	for _, name := range names {
+		v := cPkeyVecMap[name]
+		switch v.Typ.Oid {
+		case types.T_bool:
+			s := v.Col.([]bool)
+			for i, b := range s {
+				if !nulls.Contains(v.Nsp, uint64(i)) {
+					ps[i].EncodeBool(b)
+				} else {
+					nulls.Add(bat.Vecs[len(bat.Vecs)-1].Nsp, uint64(i))
+				}
+			}
+		case types.T_int8:
+			s := v.Col.([]int8)
+			for i, b := range s {
+				if !nulls.Contains(v.Nsp, uint64(i)) {
+					ps[i].EncodeInt8(b)
+				} else {
+					nulls.Add(bat.Vecs[len(bat.Vecs)-1].Nsp, uint64(i))
+				}
+			}
+		case types.T_int16:
+			s := v.Col.([]int16)
+			for i, b := range s {
+				if !nulls.Contains(v.Nsp, uint64(i)) {
+					ps[i].EncodeInt16(b)
```

Comment:
Why not just deal with it like follow way:
```sql
data := unsafe.Slice((*byte)(vector.GetPtrAt(vec, 0)), count*sz)
for i := 0; i < count; i++ {
       keys[i] = append(keys[i], data[i*sz:(i+1)*sz]...)                         
}

Reply:
There is a lot of unnecessary manipulation in the current practice.

Reply:
The fdb approach is order-preserving, and it may be necessary to introduce some overhead this way.